lib/modules.nix

```diff
@@ -782,11 +783,22 @@ rec {
     defsFinal = defsFinal'.values;
 
     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
-    mergedValue =
+    mergedValue = _mergedValueWithContext "unspecified";
+
+    # internal, unstable interface, only used locally
+    _mergedValueWithContext = parentTypeName:
       if isDefined then
         if all (def: type.check def.value) defsFinal then type.merge loc defsFinal
         else let allInvalid = filter (def: ! type.check def.value) defsFinal;
-        in throw "A definition for option `${showOption loc}' is not of type `${type.description}'. Definition values:${showDefs allInvalid}"
+        in if parentTypeName == "listOf" && type.name != "functionTo" && any (def: isFunction def.value) defsFinal
```

Comment:
`type.name != "functionTo` isn't general; not supporting `either`, among other things. How about

```nix
type.check (x:
  # We want to know whether functions are ok inside the list.
  # To determine this, we make use of the fact that a check function must only
  # evaluate to weak head normal form. This is already a requirement for making
  # option values as lazy as they need to be.
  # However, if `check` does not satisfy this rule, we don't want to trigger any
  # unexpected errors here that may originate from the weird `check` function,
  # so we make sure that any use of the body returns an appropriate error message
  # that isn't related to our little side quest of figuring out whether functions are ok.
  # This doesn't work for `check` errors resulting from functionArgs processing, but that's
  # ok, because the `check` contract is that it doesn't throw in the first place.
  throw "A definition for option `${showOption loc}' is not of type `${type.description}'. Definition values:${showDefs allInvalid}"
  )
```

The throw expression matches the original throw that is still used by `mergedValue`, so you can DRY that.

Reply:
`any (def: isFunction def.value) defsFinal`

I suppose you'd only have to check `allInvalid`?

Reply:
Note that in the current implementation `mergedValue` just calls `_mergedValueWithContext` underneath, so there's no repetition.

Regarding checking for function types, how about just not doing that. Let's just always throw this extra hint when there's a function element that didn't match the type. If the type is `functionTo` this can't happen anyways (since `check` is just `isFunction`). The only case where this is a bit misleading is when you have another custom function type whose check does some `functionArgs` processing or so. But even then, the hint could still be relevant, e.g. see this hypothetical code:

```nix
{ lib, pkgs, ... }: {
  options.fetchers = lib.mkOption {
    type = lib.types.addCheck
      (lib.types.functionTo lib.types.package)
      (f: lib.functionArgs f ? sha256);
  };
  fetchers = with pkgs; [
    fetchurl.override (old: { cacert = null; })    
    fetchzip
  ];
}
```

This shows how when there's a type mismatch for a function definition, we can't differentiate between whether the function is meant to be applied, or whether a wrongly-typed function was passed.

Reply:
> `any (def: isFunction def.value) defsFinal`
> 
> I suppose you'd only have to check `allInvalid`?

Ah yes